tutorials/Tutorial 1 - Training and Evaluation of Logistic Regression on Encrypted Data.ipynb

```diff
@@ -0,0 +1,758 @@
+    "    print(f\"Evaluated test_set of {len(x_test)} entries in {int(t_end - t_start)} seconds\")\n",
+    "    print(f\"Accuracy: {correct}/{len(x_test)} = {correct / len(x_test)}\")\n",
+    "    return correct / len(x_test)\n",
+    "    \n",
+    "\n",
+    "encrypted_accuracy = encrypted_evaluation(eelr, enc_x_test, y_test)\n",
+    "diff_accuracy = plain_accuracy - encrypted_accuracy\n",
+    "print(f\"Difference between plain and encrypted accuracies: {diff_accuracy}\")\n",
+    "if diff_accuracy < 0:\n",
+    "    print(\"Oh! We got a better accuracy on the encrypted test-set! The noise was on our side...\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We saw that evaluating on the encrypted test-set doesn't affect that much the accuracy, I've even seen examples where the encrypted evaluation performs better."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Training an Encrypted Logistic Regression Model on Encrypted Data\n",
+    "\n",
+    "In this part we will redefine a PyTorch-like model that can both forward encrypted data, as well backprobagate to update the weights and thus train the encrypted logistic regression model on encrypted data. Below are more details about the training\n",
+    "\n",
+    "#### Loss Function\n",
+    "\n",
+    "We are using the binary cross entropy loss function with regularization (more about the why of regularization will follow) where $y^{(i)}$ is the i'th expected label, $\\hat{y}^{(i)}$ is the i'th output of the logistic regression model and $\\theta$ is our n-sized weight vector.\n",
+    "\n",
+    "$$Loss(\\theta) = - \\frac{1}{m} \\sum_{i=1}^m [y^{(i)} log(\\hat{y}^{(i)}) + (1 - y^{(i)}) log (1 - \\hat{y}^{(i)})] + \\frac{\\lambda}{2m} \\sum_{j=1}^n \\theta_j^2$$\n",
+    "\n",
+    "#### Parameters Update\n",
+    "\n",
+    "For updating the parameter, the usual rule is as follows, where $x^{(i)}$ is the i'th input data\n",
+    "\n",
+    "$$\\theta_j = \\theta_j - \\alpha \\; [ \\frac{1}{m} \\sum_{i=1}^m (\\hat{y}^{(i)} - y^{(i)}) x^{(i)} + \\frac{\\lambda}{m} \\theta_j]$$\n",
+    "\n",
+    "However, due to homomorphic encryption constraint, we preferred to use an $\\alpha = 1$ to reduce a multiplication and set $\\frac{\\lambda}{m} = 0.05$ which gets us to the following update rule\n",
+    "\n",
+    "$$\\theta_j = \\theta_j - [ \\frac{1}{m} \\sum_{i=1}^m (\\hat{y}^{(i)} - y^{(i)}) x^{(i)} + 0.05 \\theta_j]$$\n",
+    "\n",
+    "#### Sigmoid Approximation\n",
+    "\n",
+    "Since we can't simply compute sigmoid on encrypted data, we need to approximate it using a low degree polynomial, the lowest degree the better, as we aim to perform as little multiplications as possible, to be able to use smaller parameters and thus optimize computation. This tutorial uses a degree 3 polynomial from https://eprint.iacr.org/2018/462.pdf which approximates the sigmoid function in the range $[-5,5]$\n",
```

Comment:
```suggestion
    "Since we can't simply compute sigmoid on encrypted data, we need to approximate it using a low degree polynomial, the lowest degree the better, as we aim to perform as little multiplications as possible, to be able to use smaller parameters and thus optimize computation. This tutorial uses a degree 1 polynomial from https://eprint.iacr.org/2018/462.pdf which approximates the sigmoid function in the range $[-5,5]$\n",
```

Reply:
I actually switched back to degree 3

Reply:
So, the formula bellow should be updated